### PR TITLE
Saner build

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: generate_version_header.sh [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #!/bin/bash
+TEMPORARY_FILE="/tmp/$RANDOM"
```
Comment (pleroy) [inline generate_version_header.sh]: `${RANDOM}`

It's good practice to always use `${}` when referencing variables.

File: geometry/permutation_test.cpp [hunk 2]
```diff
@@ -1,11 +1,12 @@
 ﻿
-#include "geometry/permutation.hpp"
-
 #include <vector>
 
 #include "geometry/frame.hpp"
 #include "geometry/identity.hpp"
 #include "geometry/orthogonal_map.hpp"
+// TODO(egg): this should be at the top, but there's a circular dependency due
+// to compatibility test.
```
Comment (pleroy) [inline geometry/permutation_test.cpp]: I'm not sure what this means, but should we have include guards to deal with this oddity?

File: install_deps.sh [hunk 3]
```diff
@@ -57,3 +57,7 @@ git clone "https://github.com/mockingbirdnest/Optional.git"
 
 # TODO(egg): This probably needs to be compiled
 git clone "https://github.com/mockingbirdnest/benchmark"
+pushd benchmark
+cmake -DCMAKE_C_COMPILER:FILEPATH=`which clang` -DCMAKE_CXX_COMPILER:FILEPATH=`which clang++` -DCMAKE_C_FLAGS="$C_FLAGS" -DCMAKE_CXX_FLAGS="$CXX_FLAGS" -DCMAKE_LD_FLAGS="$LD_FLAGS"
```
Comment (pleroy) [inline install_deps.sh]: `${C_FLAGS}` etc.

File: Makefile [hunk 4]
```diff
@@ -1,43 +1,59 @@
...
 
-GENERATED_SOURCES := journal/profiles.generated.h \
-	journal/profiles.generated.cc \
-	journal/player.generated.cc \
+OBJ_DIRECTORY := obj/
+
+BIN_DIRECTORY := bin/
+TOOLS_BIN     := $(BIN_DIRECTORY)tools
+
+GMOCK_TRANSLATION_UNITS :=                     \
```
Comment (pleroy) [inline Makefile]: Misaligned `\`.

File: Makefile [hunk 5]
```diff
@@ -1,43 +1,59 @@
...
+	$(DEP_DIR)googlemock/src/gmock-all.cc  \
+	$(DEP_DIR)googlemock/src/gmock_main.cc \
+	$(DEP_DIR)googletest/src/gtest-all.cc
+
+VERSION_HEADER := base/version.generated.h
+
+GENERATED_PROFILES :=                    \
```
Comment (pleroy) [inline Makefile]: Same.

File: Makefile [hunk 6]
```diff
@@ -1,43 +1,59 @@
...
-INCLUDES := -I. -I$(DEP_DIR)/glog/src -I$(DEP_DIR)/protobuf/src -I$(DEP_DIR)/benchmark/include -I$(DEP_DIR)/Optional $(TEST_INCLUDES)
-SHARED_ARGS := -std=c++14 -stdlib=libc++ -O3 -g -fPIC -fexceptions -ferror-limit=0 -fno-omit-frame-pointer -Wall -Wpedantic \
+ADAPTER_BUILD_DIR := ksp_plugin_adapter/obj/
+ADAPTER_CONFIGURATION := Release
+FINAL_PRODUCTS_DIR := Release/
+ADAPTER := $(ADAPTER_BUILD_DIR)$(ADAPTER_CONFIGURATION)/ksp_plugin_adapter.dll
+PLUGIN_DIRECTORY := $(FINAL_PRODUCTS_DIR)GameData/Principia/Linux64/
+
+TEST_LIBS := $(DEP_DIR)benchmark/src/libbenchmark.a
+LIBS      := $(DEP_DIR)/protobuf/src/.libs/libprotobuf.a $(DEP_DIR)/glog/.libs/libglog.a -lpthread -lc++ -lc++abi
```
Comment (pleroy) [inline Makefile]: Align all the `:=` in this block.

File: Makefile [hunk 7]
```diff
@@ -1,43 +1,59 @@
...
+TEST_INCLUDES := -I$(DEP_DIR)googlemock/include -I$(DEP_DIR)googletest/include -I$(DEP_DIR)googlemock/ -I$(DEP_DIR)googletest/ -I$(DEP_DIR)benchmark/include 
+INCLUDES := -I. -I$(DEP_DIR)glog/src -I$(DEP_DIR)protobuf/src -I$(DEP_DIR)Optional -I$(DEP_DIR)eggsperimental_filesystem/
+SHARED_ARGS := -std=c++14 -stdlib=libc++ -O3 -g -fPIC -fexceptions -ferror-limit=1 -fno-omit-frame-pointer -Wall -Wpedantic \
 	-DPROJECT_DIR='std::experimental::filesystem::path("$(PROJECT_DIR)")'\
```
Comment (pleroy) [inline Makefile]: Indent below `-std`.

File: Makefile [hunk 8]
```diff
@@ -58,118 +74,213 @@ ifeq ($(UNAME_S),Darwin)
...
+	sed 's!.*\.o[ :]*!$(OBJ_DIRECTORY)$*.o $@ : !g' < $@.temp > $@
+	rm -f $@.temp
+
+$(TEST_OR_MOCK_DEPENDENCIES): $(BUILD_DIRECTORY)%.d: %.cpp | $(PROTO_HEADERS) $(VERSION_HEADER)
+	@mkdir -p $(@D)
+	$(CXX) -M $(COMPILER_OPTIONS) $(TEST_INCLUDES) $< > $@.temp
+	sed 's!.*\.o[ :]*!$(OBJ_DIRECTORY)$*.o $@ : !g' < $@.temp > $@
+	rm -f $@.temp
+
+ifneq ($(MAKECMDGOALS),clean)
```
Comment (pleroy) [inline Makefile]: Space after comma?

File: Makefile [hunk 9]
```diff
@@ -58,118 +74,213 @@ ifeq ($(UNAME_S),Darwin)
...
+
+TEST_BINS                            := $(addprefix $(BIN_DIRECTORY), $(TEST_TRANSLATION_UNITS:.cpp=))
+PACKAGE_TEST_BINS                    := $(addprefix $(BIN_DIRECTORY), $(addsuffix test, $(sort $(dir $(TEST_TRANSLATION_UNITS)))))
+PLUGIN_DEPENDENT_TEST_BINS           := $(filter bin/ksp_plugin_test/% bin/journal/%, $(TEST_BINS))
+PLUGIN_DEPENDENT_PACKAGE_TEST_BINS   := $(filter bin/ksp_plugin_test/% bin/journal/%, $(PACKAGE_TEST_BINS))
+PLUGIN_INDEPENDENT_TEST_BINS         := $(filter-out $(PLUGIN_DEPENDENT_TEST_BINS), $(TEST_BINS))
+PLUGIN_INDEPENDENT_PACKAGE_TEST_BINS := $(filter-out $(PLUGIN_DEPENDENT_PACKAGE_TEST_BINS), $(PACKAGE_TEST_BINS))
+PRINCIPIA_TEST_BIN                   := $(BIN_DIRECTORY)test
+
+$(PLUGIN_INDEPENDENT_PACKAGE_TEST_BINS) : $(BIN_DIRECTORY)%test : $$(filter $(OBJ_DIRECTORY)%$$(PERCENT), $(TEST_OR_MOCK_OBJECTS)) $(GMOCK_OBJECTS) $(PROTO_OBJECTS) $(BASE_LIB_OBJECTS)
```
Comment (pleroy) [inline Makefile]: Can we split this looong line?

File: Makefile [hunk 10]
```diff
@@ -58,118 +74,213 @@ ifeq ($(UNAME_S),Darwin)
...
+each_test : $(TEST_TARGETS)
+each_package_test : $(PACKAGE_TEST_TARGETS)
+tidy : $(TIDY_TARGETS)
+
+.PHONY: all tools adapter plugin each_test test release clean normalize_bom tidy $(TIDY_TARGETS) $(TEST_TARGETS) $(PACKAGE_TEST_TARGETS)
+.PRECIOUS: %.o $(PROTO_HEADERS) $(PROTO_TRANSLATION_UNITS)
+.DEFAULT_GOAL := all
+.SUFFIXES:
+
+an_apple_pie : the_universe
```
Comment (pleroy) [inline Makefile]: ?

File: geometry/permutation_test.cpp (hunk 2 above)
Comment (eggrobin) [inline geometry/permutation_test.cpp]: If we actually intended to keep that, perhaps; but forward-declares and include guards to break cycles get really messy, and we want to kill compatibility code soon.

File: generate_version_header.sh (hunk 1 above)
Comment (eggrobin) [inline generate_version_header.sh]: Done.

File: install_deps.sh (hunk 3 above)
Comment (eggrobin) [inline install_deps.sh]: Done.

File: Makefile (hunk 4 above)
Comment (eggrobin) [inline Makefile]: oh. It's tab-indented because make. Let's put a single space before the first `\` and align the others.

File: Makefile (hunk 5 above)
Comment (eggrobin) [inline Makefile]: same.

File: Makefile (hunk 6 above)
Comment (eggrobin) [inline Makefile]: Done.

File: Makefile (hunk 7 above)
Comment (eggrobin) [inline Makefile]: make. tabs.

File: Makefile (hunk 8 above)
Comment (eggrobin) [inline Makefile]: Done.

**Pull request merged**